Enforce Node Allocatable via cgroups

vishh opened the pull request:
This PR enforces node allocatable across all pods using a top level cgroup as described in https://github.com/kubernetes/community/pull/348

This PR also provides an option to enforce `kubeReserved` and `systemReserved` on user specified cgroups. 

This PR will by default make kubelet create top level cgroups even if `kubeReserved` and `systemReserved` is not specified and hence `Allocatable = Capacity`.

```release-note
New Kubelet flag `--enforce-node-allocatable` with a default value of `pods` is added which will make kubelet create a top level cgroup for all pods to enforce Node Allocatable. Optionally, `system-reserved` & `kube-reserved` values can also be specified separated by comma to enforce node allocatable on cgroups specified via `--system-reserved-cgroup` & `--kube-reserved-cgroup` respectively. Note the default value of the latter flags are "".
This feature requires a **Node Drain** prior to upgrade failing which pods will be restarted if possible or terminated if they have a `RestartNever` policy.
```

cc @kubernetes/sig-node-pr-reviews @kubernetes/sig-node-feature-requests 

TODO:

- [x] Adjust effective Node Allocatable to subtract hard eviction thresholds
- [x] Add unit tests
- [x] Complete pending e2e tests
- [x] Manual testing
- [x] Get the proposal merged

@dashpole is working on adding support for evictions for enforcing Node allocatable more gracefully. That work will show up in a subsequent PR for v1.6

vishh committed:
adding kubelet flags for node allocatable phase 2

Signed-off-by: Vishnu Kannan <vishnuk@google.com>

vishh committed:
add support for node allocatable phase 2 to kubelet

Signed-off-by: Vishnu Kannan <vishnuk@google.com>

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41234)
<!-- Reviewable:end -->

vishh commented:
@derekwaynecarr @dashpole @sjenning Node e2e (finally) passes for this PR. I'd appreciate a quick review since @dashpole intends to add evictions on top of this.

derekwaynecarr commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -360,13 +375,20 @@ func (cm *containerManagerImpl) setupNode() error {
 	}
 
+	// Enforce Node Allocatable (if required)
+	if err := enforceNodeAllocatableCgroups(cm.NodeConfig.NodeAllocatableConfig,
```
fwiw, i like the model @sjenning had created where we end up with a localized interface per tier in hierarchy.  so he had a qos container manager interface that handled things at that tier, i could see us doing the same for the node allocatable tier.  makes it clear how we can keep any update logic localized.  not a must do for this pr or anything.

derekwaynecarr replied:
the other advantage i like in the model he has it has a periodic checkup pattern where as in this pr, we seem to set and forget the resource settings.

vishh replied:
I agree. I like that model too. We can have policies/methods for each tier and each controller.

I'm happy to refactor this PR later. Also, I added retries for `kubepods` cgroup just to be safe.

derekwaynecarr commented on pkg/kubelet/cm/node_allocatable.go:
```diff
@@ -0,0 +1,164 @@
+)
+
+const (
+	defaultNodeAllocatableCgroupName = "/kubepods"
```
i dont think this name should be fully qualified as we join it with the root container.

vishh replied:
It seems to be working, but yeah it is misleading. Updating it

derekwaynecarr commented on pkg/kubelet/cm/node_allocatable.go:
```diff
@@ -0,0 +1,164 @@
+
+func createNodeAllocatableCgroups(nc NodeAllocatableConfig, nodeAllocatable v1.ResourceList, cgroupManager CgroupManager) error {
+	cgroupConfig := &CgroupConfig{
+		Name: CgroupName(defaultNodeAllocatableCgroupName),
```
this isnt right if i set cgroupRoot to a value that is not equal to /

vishh replied:
Good catch. Fixing the logic.

derekwaynecarr commented on pkg/kubelet/cm/node_allocatable.go:
```diff
@@ -0,0 +1,164 @@
+
+const (
+	defaultNodeAllocatableCgroupName = "/kubepods"
+	nodeAllocatableEnforcementKey    = "pods"
```
yeah, i think these keys should be defined somewhere else so we can validate?

derekwaynecarr replied:
right now, we appear to just silently fail on "podsa" or "pod"

vishh replied:
That's true. We do not validate. I will add validation logic to `cmd/kubelet` package.

derekwaynecarr commented on pkg/kubelet/cm/node_allocatable.go:
```diff
@@ -0,0 +1,164 @@
+	glog.V(4).Infof("Attempting to enforce Node Allocatable with config: %+v", nc)
+	glog.V(4).Infof("Node Allocatable resources: %+v", nodeAllocatable)
+	// Create top level cgroups for all pods if necessary.
+	if nc.EnforceNodeAllocatable.Has(nodeAllocatableEnforcementKey) {
```
recording here so i don't forget, but do we prevent users from reserving more resource than capacity anywhere?  seems like a validation we should be able to do that system-reserved + kube-reserved < capacity, just reminding myself to check...

vishh replied:
I'm adding logic to verify this.

derekwaynecarr commented on pkg/kubelet/cm/node_allocatable.go:
```diff
@@ -0,0 +1,164 @@
+		}
+	}
+	// Now apply kube reserved and system reserved limits if required.
+	if nc.EnforceNodeAllocatable.Has(systemReservedEnforcementKey) {
```
i think we should verify the following somewhere:

systemReservedCgroupName != /
kubeReservedCgroupName != /

also need to verify that they are not equal to driver converted name for kubepods.

we should also verify they are in different branches of the hierarchy.

derekwaynecarr commented on pkg/kubelet/cm/node_allocatable.go:
```diff
@@ -0,0 +1,164 @@
+	return nil
+}
+
+func getCgroupConfig(rl v1.ResourceList) *ResourceConfig {
```
nit: i think this should make clear its a cgroup config for guaranteed QoS since there could be another method laying around that gets a cgroup config that takes a resource requirements spec.

derekwaynecarr commented on pkg/kubelet/cm/node_allocatable.go:
```diff
@@ -0,0 +1,164 @@
+	return &rc
+}
+
+func (cm *containerManagerImpl) getNodeAllocatableInternal(includeHardEviction bool) v1.ResourceList {
```
is there a test case for this?  i need to pull and physically test your pr to make sure it does what i think it is doing.

derekwaynecarr replied:
i thought we need to handle two cases:

1. allocatable reported to scheduler which was capacity - reservations
2. allocatable enforced via cgroup which was min(capacity - reservations + reservation, capacity)

derekwaynecarr replied:
after more thought, i think this is right.

so for the 32Gi, 2 gi reserve, 100Mi threshold, this would given 29.9Gi allocatable, and 30Gi enforced.  sorry, its late on a friday and this is complicated ;-)

i think the issue is its usage of `includeHardEviction`..

it would be clearer if called something like: getNodeAllocatableFor(scheduling | enforcement)

dashpole replied:
nit: I would love a comment explaining the purpose of this function.

vishh replied:
Renaming sounds useful. I will make the changes. Unit tests do exist.

vishh replied:
Done

derekwaynecarr commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -360,13 +375,20 @@ func (cm *containerManagerImpl) setupNode() error {
 
+	// Enforce Node Allocatable (if required)
+	if err := enforceNodeAllocatableCgroups(cm.NodeConfig.NodeAllocatableConfig,
+		cm.getNodeAllocatableInternal(false), // ignore eviction thresholds.
```
i thought kubepods needs to have allocatable = capacity - reserved + thresholds enforced.  right now, this appears to not give the buffer discussed.

derekwaynecarr commented on pkg/kubelet/cm/node_allocatable_test.go:
```diff
@@ -0,0 +1,114 @@
+			NodeConfig: nc,
+			capacity:   tc.capacity,
+		}
+		for k, v := range cm.GetNodeAllocatable() {
```
this is confusing me.  see my earlier comment on the internal version of the method, but this doesn't appear correct, and at minimum is not testing both patterns where we pass true or false to the operation.

derekwaynecarr replied:
after further reflection, i think we just need a test that tests the enfored values and a test that verifies the reported to scheduler value so its clearer.  i think this is correct for the as reported to scheduler case.

vishh replied:
Ack. I can split up the tests accordingly tomorrow.

derekwaynecarr commented on pkg/kubelet/cm/node_allocatable_test.go:
```diff
@@ -0,0 +1,114 @@
+		hardThreshold  evictionapi.ThresholdValue
+	}{
+		{
+			kubeReserved:   getResourceList("100m", "100Mi"),
```
add a case for the 0 expected resources scenario when thresholds cause us to exceed capacity.

vishh replied:
Adding a check in code and a test case

derekwaynecarr commented on test/e2e_node/container_manager_test.go:
```diff
@@ -70,9 +73,78 @@ func validateOOMScoreAdjSettingIsInRange(pid int, expectedMinOOMScoreAdj, expect
 var _ = framework.KubeDescribe("Kubelet Container Manager [Serial]", func() {
 	f := framework.NewDefaultFramework("kubelet-container-manager")
 
+	Describe("Validate Node Allocatable", func() {
```
@sjenning -- this seems like a pattern you could follow for your work in e2e

derekwaynecarr reviewed (changes requested):
A number of comments and requests for changes throughout.  I was confused on the getNodeAllocatable stuff, so if we can name or structure things a little differently, I will be less confused when we look again at the code.

dashpole commented on pkg/kubelet/cm/node_allocatable.go:
```diff
@@ -0,0 +1,164 @@
+}
+
+// GetNodeAllocatable returns amount of compute resource available for pods.
+func (cm *containerManagerImpl) GetNodeAllocatable() v1.ResourceList {
```
Since this is an eviction threshold, I would like to see the word threshold in this function name

dashpole replied:
Or on second thought, GetNodeAllocatableResources() would also work.

vishh replied:
This is the real `NodeAllocatable` from system and users standpoint.
The internal one is just meant for enforcing the real NodeAllocatable within kubelet.
If we were to rename, I'd rename the internal one.

dashpole commented on pkg/kubelet/cm/node_allocatable.go:
```diff
@@ -0,0 +1,164 @@
+	return nil
+}
+
+func enforceExistingCgroup(cgroupManager CgroupManager, cName string, rl v1.ResourceList) error {
```
nit: comment explaining function

vishh replied:
Done

dashpole reviewed (changes requested):
Took me a while, but I think I understand the changes.  Other than @derekwaynecarr's comments about validation, I found the allocatable code a bit difficult to read.  I think we need to make sure that everything is either a "threshold", "reservation", "limit", or "resource", since an "allocatable" doesn't have a concrete meaning.

vishh commented:
Thanks for the reviews. I will address the comments tomorrow. Meanwhile, I fixed the builds (hopefully) and also added a flag to optionally ignore eviction thresholds from node allocatable.

vishh commented on cmd/kubelet/app/options/options.go:
```diff
@@ -262,4 +261,12 @@ func (s *KubeletServer) AddFlags(fs *pflag.FlagSet) {
+	fs.Var(&s.SystemReserved, "system-reserved", "A set of ResourceName=ResourceQuantity (e.g. cpu=200m,memory=150G) pairs that describe resources reserved for non-kubernetes components. Currently only cpu and memory are supported. See http://kubernetes.io/docs/user-guide/compute-resources for more detail. [default=none]")
+	fs.Var(&s.KubeReserved, "kube-reserved", "A set of ResourceName=ResourceQuantity (e.g. cpu=200m,memory=150G) pairs that describe resources reserved for kubernetes system components. Currently only cpu and memory are supported. See http://kubernetes.io/docs/user-guide/compute-resources for more detail. [default=none]")
+
+	fs.StringSliceVar(&s.EnforceNodeAllocatable, "enforce-node-allocatable", s.EnforceNodeAllocatable, "A comma separated list of levels of node allocatable enforcement to be enforced by kubelet. Acceptible options are 'pods', 'system-reserved' & 'kube-reserved'. If the latter two options are specified, '--system-reserved-cgroup' & '--kube-reserved-cgroup' must also be set respectively. See https:// ...
```
It should render the defaults, but I suspect the default string value is meant for automated docs.

vishh commented on cmd/kubelet/app/server.go:
```diff
@@ -305,6 +308,17 @@ func initConfigz(kc *componentconfig.KubeletConfiguration) (*configz.Config, err
+	if !s.CgroupsPerQOS && len(s.EnforceNodeAllocatable) > 0 {
+		return fmt.Errorf("Node Allocatable enforcement is not supported unless Cgroups Per QOS feature is turned on")
+	}
+	if s.SystemCgroups != "" && s.CgroupRoot == "" {
```
It is still being used in the old debian based distro that GKE uses. We can deprecate it once GKE stops using that distro.

derekwaynecarr replied:
sgtm - not needed by this pr, but we should probably put a big warning around those flags that you should not use them if your init system already places your stuff in a cgroup.  maybe even error out completely if they are used and you run systemd.

vishh commented on pkg/apis/componentconfig/v1alpha1/defaults.go:
```diff
@@ -48,7 +48,12 @@ const (
+	zeroDuration = metav1.Duration{}
+	// Refer to [Node Allocatable](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node-allocatable.md) doc for more information.
+	// TODO: Set the default to "pods" once cgroups per qos is turned on by default.
+	defaultNodeAllocatableEnforcement = []string{}
```
Ack

vishh commented:
Fixed most of the comments excepting splitting of unit tests and nits.

derekwaynecarr commented:
@vishh -- i will take another pass in the morning.

dashpole commented on pkg/kubelet/cm/node_allocatable.go:
```diff
@@ -0,0 +1,216 @@
+		// If Node Allocatable is enforced on a node that has not been drained or is updated on an existing node to a lower value,
+		// existing memory usage across pods might be higher that current Node Allocatable Memory Limits.
+		// Pod Evictions are expected to bring down memory usage to below Node Allocatable limits.
+		// Until evictions happen retry cgroup updates.
```
Not sure if this is completed given the TODO below, but should this function return after a successful update?
And do we have a way to test this?

vishh replied:
There isn't a test for the retries yet. There is a test for verifying basic enforcement though. A retry test can be added once support for evictions are added (by you :) )

derekwaynecarr replied:
i have to admit i am not the biggest fan of anonymous inner go funcs ;-)

wont this spit out an event every minute forever each time that it correctly updates the allocatable limit across pods?  this appears to run forever... i am not sure i need that many events ;-)

vishh commented on cmd/kubelet/app/options/options.go:
```diff
@@ -224,8 +224,7 @@ func (s *KubeletServer) AddFlags(fs *pflag.FlagSet) {
 	fs.Int64Var(&s.MaxOpenFiles, "max-open-files", s.MaxOpenFiles, "Number of files that can be opened by Kubelet process. [default=1000000]")
-	fs.Var(&s.SystemReserved, "system-reserved", "A set of ResourceName=ResourceQuantity (e.g. cpu=200m,memory=150G) pairs that describe resources reserved for non-kubernetes components. Currently only cpu and memory are supported. See http://kubernetes.io/docs/user-guide/compute-resources for more detail. [default=none]")
-	fs.Var(&s.KubeReserved, "kube-reserved", "A set of ResourceName=ResourceQuantity (e.g. cpu=200m,memory=150G) pairs that describe resources reserved for kubernetes system components. Currently only cpu and memory are supported. See http://kubernetes.io/docs/user-guide/compute-resources for more detail. [default=none]")
+
```
ack

vishh commented on cmd/kubelet/app/options/options.go:
```diff
@@ -262,4 +261,12 @@ func (s *KubeletServer) AddFlags(fs *pflag.FlagSet) {
+	// Node Allocatable Flags
+	fs.Var(&s.SystemReserved, "system-reserved", "A set of ResourceName=ResourceQuantity (e.g. cpu=200m,memory=150G) pairs that describe resources reserved for non-kubernetes components. Currently only cpu and memory are supported. See http://kubernetes.io/docs/user-guide/compute-resources for more detail. [default=none]")
+	fs.Var(&s.KubeReserved, "kube-reserved", "A set of ResourceName=ResourceQuantity (e.g. cpu=200m,memory=150G) pairs that describe resources reserved for kubernetes system components. Currently only cpu and memory are supported. See http://kubernetes.io/docs/user-guide/compute-resources for more detail. [default=none]")
+
```
ack

derekwaynecarr commented:
@k8s-bot test this

derekwaynecarr commented:
I would like https://github.com/kubernetes/kubernetes/pull/41349 to merge before this PR.

derekwaynecarr commented on cmd/kubelet/app/options/options.go:
```diff
@@ -262,4 +260,12 @@ func (s *KubeletServer) AddFlags(fs *pflag.FlagSet) {
+	fs.StringSliceVar(&s.EnforceNodeAllocatable, "enforce-node-allocatable", s.EnforceNodeAllocatable, "A comma separated list of levels of node allocatable enforcement to be enforced by kubelet. Acceptible options are 'pods', 'system-reserved' & 'kube-reserved'. If the latter two options are specified, '--system-reserved-cgroup' & '--kube-reserved-cgroup' must also be set respectively. See https:// ...
+	fs.StringVar(&s.SystemReservedCgroup, "system-reserved-cgroup", s.SystemReservedCgroup, "Absolute name of the top level cgroup that is used to manage non-kubernetes components for which compute resources were reserved via '--system-reserved' flag. Ex. '/system-reserved'. [default='']")
+	fs.StringVar(&s.KubeReservedCgroup, "kube-reserved-cgroup", s.KubeReservedCgroup, "Absolute name of the top level cgroup that is used to manage kubernetes components for which compute resources were reserved via '--kube-reserved' flag. Ex. '/kube-reserved'. [default='']")
+	fs.BoolVar(&s.ExperimentalNodeAllocatableIgnoreEvictionThreshold, "experimental-node-allocatable-ignore-eviction-threshold", s.ExperimentalNodeAllocatableIgnoreEvictionThreshold, "When set to 'true', Hard Eviction Thresholds will be ignored while calculating Node Allocatable. See https://github.com/kubernetes/community/blob/master/contributors/design-proposals/node-allocatable.md for more detail ...
```
this might hold the record for all time longest flag name.

vishh replied:
:) Any suggestions on an alternative name?

derekwaynecarr commented on cmd/kubelet/app/server.go:
```diff
@@ -465,6 +493,14 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.KubeletDeps) (err error) {
+				NodeAllocatableConfig: cm.NodeAllocatableConfig{
+					KubeReservedCgroupName:   s.KubeReservedCgroup,
+					SystemReservedCgroupName: s.SystemReservedCgroup,
+					EnforceNodeAllocatable:   sets.NewString(s.EnforceNodeAllocatable...),
```
ack that this is now validated.

derekwaynecarr commented on cmd/kubelet/app/server.go:
```diff
@@ -681,15 +747,16 @@ func RunKubelet(kubeCfg *componentconfig.KubeletConfiguration, kubeDeps *kubelet
-	} else {
-		glog.Warning("No api server defined - no events will be sent to API server.")
+	if kubeDeps.Recorder == nil {
+		eventBroadcaster := record.NewBroadcaster()
```
nit: this is duplicate code, we could pull it out into a `makeRecorder` function since it is now present twice.

dashpole replied:
done.

vishh replied:
Done.

derekwaynecarr commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -204,7 +215,15 @@ func NewContainerManager(mountUtil mount.Interface, cadvisorInterface cadvisor.I
 			"In the meantime, you can opt-in to making this a fatal error by enabling --experimental-fail-swap-on.")
 	}
+	var capacity = v1.ResourceList{}
+	if info, err := cadvisorInterface.MachineInfo(); err == nil {
```
is cadvisor actually guaranteed to be ready at this point with this information?  this is called prior to cadvisor.Start() since its an injected dependency...

vishh replied:
Yes, it is safe. Adding a comment explaining it. Good question though :)

derekwaynecarr commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -204,7 +215,15 @@ func NewContainerManager(mountUtil mount.Interface, cadvisorInterface cadvisor.I
+		return nil, err
+	}
 
+	cgroupRoot := nodeConfig.CgroupRoot
```
this is fine in a follow-on, but I think in future support issues, it would be a great help for us to have a log function that outputs what the node config was that we were given at this point.

derekwaynecarr replied:
fyi @sjenning -- can you take the todo to make sure we have a call that basically dumps the output of node config into a log so we can handle a variety of bad support cases in the future with sane logs?

vishh replied:
I'm adding a log line for it.

derekwaynecarr commented on pkg/kubelet/cm/node_allocatable.go:
```diff
@@ -0,0 +1,216 @@
+				glog.V(4).Infof("Attempting to update Node Allocatable cgroup with %d cpu shares and %d bytes of memory", cgroupConfig.ResourceParameters.CpuShares, cgroupConfig.ResourceParameters.Memory)
+				err := cm.cgroupManager.Update(cgroupConfig)
+				if err == nil {
+					r.Event(node, v1.EventTypeNormal, events.SuccessfulNodeAllocatableEnforcement, "Updated Node Allocatable limit across pods")
```
should there be a return here or were you looking to perpetually update periodically every minute?

dashpole replied:
I had the same question above.  If I read the code correctly, cgroupConfig is constant, and thus a successful call to `cm.cgroupManager.Update(cgroupConfig)` should mean that this loop is no longer needed.  Ill leave it as is for now, and hopefully @vishh can address this.

vishh replied:
Yeah. I was planning on breaking after a successful update. Mistake.

derekwaynecarr commented on pkg/kubelet/cm/node_allocatable.go:
```diff
@@ -0,0 +1,216 @@
+	// Now apply kube reserved and system reserved limits if required.
+	if nc.EnforceNodeAllocatable.Has(SystemReservedEnforcementKey) {
+		glog.V(2).Infof("Enforcing system reserved on cgroup %q with limits: %+v", nc.SystemReservedCgroupName, nc.SystemReserved)
+		if err := enforceExistingCgroup(cm.cgroupManager, nc.SystemReservedCgroupName, nc.SystemReserved); err != nil {
```
i find it odd we do this in a one time step, but we appear to continually attempt to update the pods cgroup...

vishh replied:
Good question. I thought about it while coding but prioritized pod level cgroup over the rest since its critical for v1.6. I'm adding retries for all of them now though.

vishh replied:
Actually, I think I recall why a retry isn't useful. System reserved and kube reserved cgroups are not managed by kubelet and so it cannot reclaim memory in the case of failures. So it doesn't make sense to retry. Adding an event makes sense though, but kubelet initialization would fail to make the error obvious.

vishh replied:
Adding an event for failures and success for system and kube reserved cgroup updates.

derekwaynecarr commented on pkg/kubelet/cm/node_allocatable.go:
```diff
@@ -0,0 +1,216 @@
+}
+
+// enforceExistingCgroup updates the limits `rl` on existing cgroup `cName` using `cgroupManager` interface.
+func enforceExistingCgroup(cgroupManager CgroupManager, cName string, rl v1.ResourceList) error {
```
i am not sure why we try once for the system groups, but forever for the pod group.

i see the comment on upgrade scenarios, but isnt the upgrade scenario valid for the system services as well?

derekwaynecarr replied:
i imagine an event would be useful for the sytem-reserved and kube-reserved sucess cases as well.

vishh replied:
Done

derekwaynecarr commented on pkg/kubelet/cm/node_allocatable.go:
```diff
@@ -0,0 +1,216 @@
+// If `forScheduling` is the purpose, then hard eviction thresholds are subtracted from capacity while computing Node Allocatable.
+// If `forEnforcement` if the purpose, then hard eviction thresholds are ignored.
+// Returns a ResourceList.
+func (cm *containerManagerImpl) getNodeAllocatableInternal(purpose int) v1.ResourceList {
```
feels like we should have an enum of some kind... the int feels a bit open ended here...

dashpole replied:
used an enum

vishh replied:
ack

derekwaynecarr reviewed (changes requested):
took another pass, had another round of questions and comments.  will work with @dashpole to close out a review on this.

vishh commented:
Sorry for the sloppy work on the recent update. I rushed through it a bit. I will fix it up now.

vishh commented:
@k8s-bot test this

vishh commented:
I got around to update this patch once again. Given the release crunch, it becoming a pain to rebase it. I'd appreciate another around of quick review @derekwaynecarr & @dashpole

vishh commented:
@k8s-bot test this

dashpole commented:
@k8s-bot test this

dashpole commented:
@k8s-bot test this

dashpole commented:
@k8s-bot test this

dashpole commented:
Errors are due to a difference between the old kubelet_node_status.go, and the new container_manager.go's calculation of capacity.  The new version omits resources such as v1.ResourcePods, and v1.ResourceGPU.  I am adding those now.

vishh commented:
Good catch @dashpole.

dashpole commented:
I changed one more thing: during validation, we rejected allocatable resources that had a 0 value, such as GPU.  I set it to accept 0 values, by removing the "=" sign.

dashpole commented:
Looks like the e2e-node cgroup test still fail.  It looks like they test the cgroup structure, so we may need to change those.

vishh commented on cmd/kubelet/app/server.go:
```diff
@@ -525,6 +526,21 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.KubeletDeps) (err error) {
+		} else {
+			capacity[v1.ResourcePods] = *resource.NewQuantity(int64(s.KubeletConfiguration.MaxPods), resource.DecimalSI)
+		}
+		capacity[v1.ResourceNvidiaGPU] = *resource.NewQuantity(int64(s.KubeletConfiguration.NvidiaGPUs), resource.DecimalSI)
```
Resources can get added by other components on a cluster too. So from a Node Allocatable standpoint, we only want to reserve some resource off of capacity.
I feel a better approach might be that of treating the list of resources that`GetNodeAllocatable()` returns as a reservation and have kubelet_node_status.go subtract that reservation from Capacity. WDYT?
The `getNodeAllocatableInternal` is still needed for enforcement purposes.

derekwaynecarr replied:
+1

derekwaynecarr commented on pkg/kubelet/cm/node_allocatable.go:
```diff
@@ -0,0 +1,234 @@
+		// existing memory usage across pods might be higher that current Node Allocatable Memory Limits.
+		// Pod Evictions are expected to bring down memory usage to below Node Allocatable limits.
+		// Until evictions happen retry cgroup updates.
+		go runUntilSuccess(
```
why not use `go wait.PollImmediateInfinite(...)`?

https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go#L233

vishh replied:
That's possible, but that would result in a larger anonymous method just like the previous iteration.

derekwaynecarr commented:
poke when e2es pass.

i still would prefer a code structure that has a separate file for each tier in the hierarchy just from a sanity standpoint.  there is a lot of stuff in container_manager_linux.go, and we could have a structure like:

+ container_manager_linux.go
+ node_container_manager
+ qos_container_manager (which 
+ pod_container_manager

just so i can quickly pivot to each tier in the hierarchy...

i want all our things to make code freeze, but i think we can look to structure this a little better post freeze?

@sjenning explored similar here: https://github.com/kubernetes/kubernetes/pull/41833

vishh commented:
+1 for improving the code structure. I added `node_allocatable.go` specifically for that purpose.

vishh commented:
I fixed the allocatable computation issue and renamed the newly added file as per https://github.com/kubernetes/kubernetes/pull/41234#issuecomment-282200160

vishh commented:
Upon reviewing #41833 I see what you meant @derekwaynecarr. I don't have cycles to create a new object for node allocatable in this PR. Let's clean it up in a subsequent PR.

dashpole commented:
`test/e2e_node/pods_container_manager_test.go:63: cannot use cgroupName (type cm.CgroupName) as type string in argument to path.Join`
`test/e2e_node/pods_container_manager_test.go:63: cannot use path.Join(defaultNodeAllocatableCgroup, cgroupName) (type string) as type cm.CgroupName in assignment`

vishh commented:
Fixed that. Now let's see what else is breaking.

vishh commented:
@derekwaynecarr @dashpole take a look at the last commit. It is attempting to prevent CPU starvation.

dashpole commented:
TEST: `[It] pod infra containers oom-score-adj should be -998 and best effort container's should be 1000`
ERROR: `spec.containers[0].name: Invalid value: "besteffort���/��\x11懊B\x01\n�\x00\x02"`

dashpole commented:
great.  hopefully this passes everything :)

vishh commented:
@derekwaynecarr @dashpole I think this PR will be good to go now. The final commit https://github.com/kubernetes/kubernetes/pull/41234/commits/d4570467f3ccaadac3802292290045580758f7d1 should fix the node e2e failures. I'm glad we have these e2es :)

Given the rebase mayhem, I'm inclined to get this PR through asap.

dashpole commented on pkg/kubelet/kubelet_node_status.go:
```diff
@@ -522,18 +522,14 @@ func (kl *Kubelet) setNodeStatusMachineInfo(node *v1.Node) {
-		if value.Sign() < 0 {
-			// Negative Allocatable resources don't make sense.
-			value.Set(0)
+		if res, exists := allocatableReservation[k]; exists {
```
I like this design.

dashpole commented on pkg/kubelet/cm/node_container_manager.go:
```diff
@@ -0,0 +1,243 @@
+// getNodeAllocatableAbsolute returns the absolute value of Node Allocatable which is primarily useful for enforcement.
+// Note that not all resources that are available on the node are included in the returned list of resources.
+// Returns a ResourceList.
+func (cm *containerManagerImpl) getNodeAllocatableAbsolute() v1.ResourceList {
```
nit: maybe getNodeAllocatableLimits?  Absolute doesn't mean much to me, since we don't have a getNodeAllocatableRelative function to compare it to.

derekwaynecarr replied:
i think we can massage names post feature freeze.

dashpole commented:
Looks good to me Vish!  Dont worry about the nit.  I can change that when I do allocatable eviction if you think it is a good idea.

vishh commented:
Yeah. I want to shift my focus a bit on a pending GPU pr. I will appreciate if we can let this PR fly by for now as long as it is correct.

derekwaynecarr commented:
There is a bug in this PR here:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/cm/pod_container_manager_linux.go#L208

The kubepods prefix is causing a bad split.

derekwaynecarr commented:
which i admit is brittle, it should have been looking at the last path part only.

derekwaynecarr commented:
i am trying to hack something on your pr that fixes it, there is a problem in that area, but i need figure a proper fix.

derekwaynecarr commented:
@vishh -- per our chat, i will post a patch to your pr this afternoon, there is a bug in `GetAllPodsFromCgroups` that must be fixed prior to this merging.

derekwaynecarr committed:
Fix get all pods from cgroups logic

derekwaynecarr commented:
@vishh -- see the pull on this branch here: https://github.com/vishh/kubernetes/pull/5

dashpole commented:
Just took a full look over this again, and this looks good to me!

vishh commented:
@dashpole thanks for the review

@derekwaynecarr I merged your patch and revamped the e2e suite completely. I wasted too much time working around ginkgo's expectations and now I have written it as a simple go program.
The e2e is now complete. It verified kube and system reserved enforcement as well.

vishh committed:
fix go vet issues

Signed-off-by: Vishnu kannan <vishnuk@google.com>

vishh committed:
update flags script

Signed-off-by: Vishnu kannan <vishnuk@google.com>

vishh commented:
@k8s-bot cri node e2e test this

vishh commented:
@k8s-bot gci gce e2e test this

vishh commented:
@k8s-bot non-cri e2e test this

vishh commented:
e2e failures seem to be unrelated. CRI node e2e failures are pointing to non-existent code issues.

derekwaynecarr commented on pkg/kubelet/cm/types.go:
```diff
@@ -25,6 +25,8 @@ import (
 	// Memory limit (in bytes).
 	Memory *int64
+	// Hierarchical Memory Accounting.
+	HierarchicalMemoryAccounting bool
```
i dont see where this is used at the moment...

vishh replied:
Oops. This change wasn't meant to be included. Removing it.

derekwaynecarr commented:
I think this PR is in good enough shape to merge now.

We can handle any naming nits post feature freeze.

/lgtm

vishh commented:
Adding lgtm again after removing unused code.

dashpole commented:
@k8s-bot test this

vishh commented:
@k8s-bot test this

derekwaynecarr commented:
to make it easier for folks rebasing on this, can you make any subsquent changes in additive commits to this pr?

vishh commented:
Ack!

On Mon, Feb 27, 2017 at 12:08 PM, Derek Carr <notifications@github.com>
wrote:

> to make it easier for folks rebasing on this, can you make any subsquent
> changes in additive commits to this pr?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/41234#issuecomment-282837050>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKB1Ull0MoKM1lWLPcC9zDFaw1lkXks5rgy0mgaJpZM4L9BCR>
> .
>

vishh commented:
@k8s-bot cri node e2e test this

vishh commented:
@k8s-bot gci gce e2e test this

derekwaynecarr commented:
marking do not merge -- it appears to be missing: https://github.com/vishh/kubernetes/pull/5

derekwaynecarr commented on pkg/kubelet/cm/container_manager_linux.go:
```diff
@@ -216,17 +237,24 @@ func NewContainerManager(mountUtil mount.Interface, cadvisorInterface cadvisor.I
+		glog.Infof("container manager verified user specified cgroup-root exists: %v", cgroupRoot)
+		// Include the the top level cgroup for enforcing node allocatable into cgroup-root.
+		// This way, all sub modules can avoid having to understand the concept of node allocatable.
+		cgroupRoot = path.Join(cgroupRoot, defaultNodeAllocatableCgroupName)
```
this should only happen if enforcement is on.  otherwise, the behavior is wrong.

vishh replied:
I thought you wanted all cgroup hierarchy changes to be tied to `cgroups-per-qos` feature? Originally, I had this the way you want it now. I think the way it is better from a rollout standpoint.
Make all cgroup changes in one shot and then apply allocatable restrictions over time.

derekwaynecarr replied:
ah, that is right.

derekwaynecarr reviewed (changes requested):
see the comments on missing commits, and how this doesnt work if enforcement is off.

derekwaynecarr reviewed (changes requested):
one more problem...

vishh commented on pkg/kubelet/eviction/helpers.go:
```diff
@@ -852,8 +845,8 @@ func hasThreshold(inputs []Threshold, item Threshold) bool {
 
-// compareThresholdValue returns true if the two thresholdValue objects are logically the same
-func compareThresholdValue(a ThresholdValue, b ThresholdValue) bool {
+// compareevictionapi.ThresholdValue returns true if the two thresholdValue objects are logically the same
```
Fixing this. Good catch.

vishh committed:
fix eviction helper function description

Signed-off-by: Vishnu kannan <vishnuk@google.com>

vishh commented:
@derekwaynecarr PTAL

derekwaynecarr commented:
pulled updates, and tried it out locally.  things look good now.

derekwaynecarr commented:
@k8s-bot gce etcd3 e2e test this

vishh commented:
Err. rebase :(

vishh commented:
Bumping the priority further up because there are about 4+ PRs that are built on top of this.

vishh commented:
adding LGTM label after rebasing.

yshngg commented on cmd/kubelet/app/server.go:
```diff
@@ -448,14 +501,30 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.KubeletDeps) (err error) {
 	}
 
+	// Setup event recorder if required.
+	makeEventRecorder(&s.KubeletConfiguration, kubeDeps, nodeName)
```
Hi @vishh , while reading the source code related **EventRecorder** in **kubelet**, I noticed that `makeEventRecorder` is called twice in the `run` function: one before `ContainerManager` initialization and once in the `RunKubelet` function. So i'm here to ask why `makeEventRecorder` is called before `ContainerManager` initialization. Thank you! 🙏
